R/metadata.R

```diff
@@ -0,0 +1,124 @@
+parse_metadata <- function(filename) {
+  data <- decor::cpp_decorations(files = filename)
+  ret <- list(name = parse_metadata_name(data),
+              class = parse_metadata_class(data),
+              param = parse_metadata_param(data))
+
+  if (is.null(ret$class)) {
+    ret$class <- parse_metadata_guess_class(readLines(filename))
+  }
+
+  if (is.null(ret$name)) {
+    ret$name <- ret$class
+  }
+
+  assert_valid_name(ret$name)
+
+  ret
+}
+
+
+parse_metadata_simple <- function(data, attribute) {
+  i <- data$decoration == attribute
+  if (!any(i)) {
+    return(NULL)
+  }
+  if (sum(i) > 1) {
+    stop(sprintf("More than one [[%s()]] attribute found %s",
+                 attribute, parse_metadata_describe(data, i)))
+  }
+  value <- data$params[[which(i)]]
+  if (length(value) != 1L) {
+    stop(sprintf("Expected [[%s()]] to have one argument %s",
+                 attribute, parse_metadata_describe(data, i)))
+  }
+  if (any(nzchar(names(value)))) {
+    stop(sprintf("Invalid format for [[%s()]] attribute %s",
+                 attribute, parse_metadata_describe(data, i)))
+  }
+  as.character(value[[1]])
+}
+
+
+parse_metadata_name <- function(data) {
+  parse_metadata_simple(data, "dust::name")
+}
+
+
+parse_metadata_class <- function(data) {
+  parse_metadata_simple(data, "dust::class")
+}
+
+
+parse_metadata_param <- function(data) {
+  i <- data$decoration == "dust::param"
+  if (!any(i)) {
+    return(NULL)
+  }
```

Comment:
What is the behaviour when the decorators are not used for a parameter? Still compiles ok, but is `coef()` uncallable/incomplete?

Reply:
yes, it keeps the current behaviour, and `coef()` etc return `NULL`